### PR TITLE
SILGen: Port reabstraction thunks to the new function type representation

### PR DESCRIPTION
**Pull request by slavapestov**
This was the last place we used AbstractionPattern::getFunctionInputType(), so a lot of old code can be removed now.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
fluent failure is the same non-deterministic associated type thing (rdar://44777661).

File: lib/SILGen/SILGenPoly.cpp [hunk 1]
```diff
@@ -892,11 +879,102 @@ namespace {
                    AnyFunctionType::CanParamArrayRef inputSubstTypes,
                    AbstractionPattern outputOrigFunctionType,
                    AnyFunctionType::CanParamArrayRef outputSubstTypes) {
+      // FIXME: Do we ever do opaque-to-opaque reabstractions?
+      assert(!inputOrigFunctionType.isTypeParameter() ||
+             !outputOrigFunctionType.isTypeParameter());
```
Comment (rjmccall) [inline lib/SILGen/SILGenPoly.cpp]: No, we don't.

**Review by rjmccall — APPROVED** [review]
Oh man, seeing the parameter-tuple abstraction patterns go away is glorious.

**Pull request merged**